From 573f0ecb29000ad70dc04fcabd9fa45744ade9f3 Mon Sep 17 00:00:00 2001
From: Kien Nguyen <kien.n.quang@gmail.com>
Date: Wed, 6 Apr 2022 09:21:32 +0900
Subject: [PATCH] Make file-name-split returns driver name as well in Windows

* lisp/files.el (file-name-split): returns driver name as well in Windows
  lisp/net/browse-url.el (browse-url-file-url): dont hexify colon character in file path for Windows (bug#54721)

---
 lisp/files.el          | 2 +-
 lisp/net/browse-url.el | 7 ++++---
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/lisp/files.el b/lisp/files.el
index a0bc5bf262..c198f18723 100644
--- a/lisp/files.el
+++ b/lisp/files.el
@@ -5091,7 +5091,7 @@ file-name-split
         ;; If there's nothing left to peel off, we're at the root and
         ;; we can stop.
         (when (and dir (equal dir filename))
-          (push "" components)
+          (push (if (string-empty-p dir) "" (substring dir 0 -1)) components)
           (setq filename nil))))
     components))
 
diff --git a/lisp/net/browse-url.el b/lisp/net/browse-url.el
index 4c348781a8..0fbf2d40f4 100644
--- a/lisp/net/browse-url.el
+++ b/lisp/net/browse-url.el
@@ -728,9 +728,10 @@ browse-url-file-url
                      browse-url-filename-alist))
       (setq file (browse-url-url-encode-chars file "[*\"()',=;?% ]"))
     ;; Encode all other file names properly.
-    (setq file (mapconcat #'url-hexify-string
-                          (file-name-split file)
-                          "/")))
+    (let ((url-unreserved-chars `(?: ,@url-unreserved-chars)))
+      (setq file (mapconcat #'url-hexify-string
+                            (file-name-split file)
+                            "/"))))
   (dolist (map browse-url-filename-alist)
     (when (and map (string-match (car map) file))
       (setq file (replace-match (cdr map) t nil file))))
-- 
2.35.1.windows.2

